fc-extract-delta/lib/staterootfilter.go

```diff
@@ -0,0 +1,80 @@
+package lib
+
+import (
+	"context"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/state"
+	cid "github.com/ipfs/go-cid"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/go-address"
+)
+
+func GetFilteredStateRoot(ctx context.Context, a api.FullNode, msg cid.Cid) (*state.StateTree, error) {
+	msgInfo, err := a.StateSearchMsg(ctx, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	ts, err := a.ChainGetTipSet(ctx, msgInfo.TipSet)
+	if err != nil {
+		return nil, err
+	}
+
+	store := apiIpldStore{ctx, a}
+	tree, err := state.LoadStateTree(&store, ts.ParentState())
+	if err != nil {
+		return nil, err
+	}
+
+	goodActors, err := GetActorsForMessage(ctx, a, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	allActors, err := a.StateListActors(ctx, msgInfo.TipSet)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, act := range allActors {
+		if _, ok := goodActors[act]; ok {
+			continue
+		}
+
+		if err := tree.DeleteActor(act); err != nil {
+			return nil, err
+		}
+	}
+
+	if 	_, err = tree.Flush(ctx); err != nil {
+		return nil, err
+	}
+
+	return tree, nil
+}
+
+func GetActorsForMessage(ctx context.Context, a api.FullNode, msg cid.Cid) (map[address.Address]struct{}, error) {
+	msgInfo, err := a.StateSearchMsg(ctx, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	trace, err := a.StateReplay(ctx, msgInfo.TipSet, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	addresses := make(map[address.Address]struct{})
+	populateFromTrace(&addresses, &trace.ExecutionTrace)
+	return addresses, nil
+}
+
+func populateFromTrace(m *map[address.Address]struct{}, trace *types.ExecutionTrace) {
```

Comment:
maps are passed by reference in golang, so no need for the pointer here: 

* https://play.golang.org/p/AWchSOsnU6b
* https://stackoverflow.com/questions/40680981/are-maps-passed-by-value-or-by-reference-in-go